Source/Dafny/Compilers/Compiler.cs

```diff
@@ -3068,14 +3068,15 @@ public class CheckHasNoAssumes_Visitor : BottomUpVisitor {
         var i = 0;
         foreach (var local in s.Locals) {
           bool hasRhs = s.Update is AssignSuchThatStmt;
+          bool willBeAssignedAnyway = s.Update is AssignOrReturnStmt;
           if (!hasRhs && s.Update is UpdateStmt u) {
             if (i < u.Rhss.Count && u.Rhss[i] is HavocRhs) {
               // there's no specific initial value
             } else {
               hasRhs = true;
             }
           }
-          TrLocalVar(local, !hasRhs, wr);
+          TrLocalVar(local, !hasRhs && !willBeAssignedAnyway, wr);
```

Comment:
Nice. Right, it makes sense to not initialize it if the type itself is not known to have a value, that's even better.

However, even if the type is known to have a value, we still don't need to initialize it if the statement is an `:-` with a right hand side.
Hence I suggest the strongest statement:

          TrLocalVar(local, !hasRhs && local.Type.KnownToHaveToAValue(false) && !willBeAssignedAnyway, wr);

Are you ok with it, or do you still prefer to do the initialization although it will be either overriden or discarded for the case of the AssignAndUpdate statement?

Reply:
I completely overlooked the meaning of `hasRHS`. I can remove `willBeAssignedAnyway` by changing the definition of `hasRHS`:

`bool hasRhs = s.Update is AssignSuchThatStmt || s.Update is AssignOrReturnStmt`